pkg/internal/serialization.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"github.com/go-openapi/jsonreference"
+)
+
+// JSONRefFromMap populates a json reference object if the map v contains a $ref key.
+func JSONRefFromMap(jsonRef *jsonreference.Ref, v map[string]interface{}) error {
+	if v == nil {
+		return nil
+	}
+	if vv, ok := v["$ref"]; ok {
+		if str, ok := vv.(string); ok {
+			ref, err := jsonreference.New(str)
+			if err != nil {
+				return err
+			}
+			*jsonRef = ref
+		}
+	}
+	return nil
+}
+
+// SanitizeExtensions sanitizes the input map such that non extension
+// keys (non x-*, X-*) keys are dropped from the map
+func SanitizeExtensions(e map[string]interface{}) {
+	for k := range e {
+		if !IsExtensionKey(k) {
+			delete(e, k)
+		}
+	}
+}
```

Comment:
OK What about this then ;-)
```suggestion
// SanitizeExtensions sanitizes the input map such that non extension
// keys (non x-*, X-*) keys are dropped from the map. Returns the new 
// modified map, or nil if the map is now empty.
func SanitizeExtensions(e map[string]interface{}) map[string]interface{} {
	for k := range e {
		if !IsExtensionKey(k) {
			delete(e, k)
		}
	}
	if len(e) == 0 {
	    e = nil
	}
	return e
}
```

And then everywhere you can do something like ...
```
p.Extensions = internal.SanitizeExtensions(x.Extensions)
```
it's a little misleading because sanitize doesn't make a new map, it modifies in place AND returns itself. I don't mind keeping it this way otherwise.

WDYT?

Reply:
Agree on the slightly misleading part so we should be explicit in the documentation but I do like how concise it makes the unmarshal code. Updated